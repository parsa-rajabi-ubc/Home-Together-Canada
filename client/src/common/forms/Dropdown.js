/**
 * @Author:     Parsa Rajabi
 * @Created:    2020.11.23
 *
 * @Description:  Re-usable Dropdown Component
 *
 */

import React, {useEffect, useState} from 'react';
import Select from 'react-select';
import propTypes from "prop-types";

const newStyling = {
    control: base => ({
            ...base,
            marginTop: 4,
            borderColor: "#e2e8f0",
            marginBottom: 16,
            paddingTop: 2,
            paddingBottom: 2,
        }
    ),
    menuPortal: base => ({...base, zIndex: 9999}),
}

function Dropdown(props) {
    const {isSearchable, name, placeholder, options, onChange, intialSelection, styling = newStyling} = props;

    const [selected, setSelected] = useState(intialSelection || "");

    // this code is run every time selected changes
    useEffect(() => {
        // this onChange function is the callback from the parent component
        onChange(selected);
        // that can be used to get the value that is inside the dropdown
    }, [selected]);

    return (
        <div>
            <Select
                isSearchable={isSearchable}
                placeholder={placeholder}
                options={options}
                value={options.find(obj => obj.label === selected)}
                onChange={(e) => setSelected(e)}
                name={name}
                menuPortalTarget={document.body}
                styles={styling}
                theme={theme => ({
                    ...theme,
                    borderRadius: 8,
                    colors: {
                        ...theme.colors,
                        neutral50: '#A0AEBF',  // Placeholder color
                    }
                })}
            />
        </div>
    );
}

Dropdown.propTypes = {
    options: propTypes.array.isRequired,
    name: propTypes.string,
    isSearchable: propTypes.bool,
    placeholder: propTypes.string,
    onChange: propTypes.func,
    intialSelection: propTypes.string,
    styling: propTypes.object
};

export default Dropdown;
