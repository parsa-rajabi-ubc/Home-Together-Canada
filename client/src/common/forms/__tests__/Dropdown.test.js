/**
 * @Author:     Parsa Rajabi
 * @Created:    2020.11.23
 *
 * @Description: Testing Dropdown Component
 *
 */
import React from 'react';
import renderer from  'react-test-renderer'
import Dropdown from "../Dropdown";

describe('Dropdown', () => {
    describe('Snapshot test', () => {
        it("should render correctly regardless of properties", () => {
            // given
            const className = 'CSS';
            const isSearchable = true;
            const onChange = jest.fn();
            const placeholder = 'placeholders';
            const options = [
                { value: 'chocolate', label: 'Chocolate' },
                { value: 'strawberry', label: 'Strawberry' },
                { value: 'vanilla', label: 'Vanilla' },
            ];
            // when
            const component = renderer.create(<Dropdown className={className} isSearchable={isSearchable} options={options} placeholder={placeholder}  onChange={onChange}/>).toJSON();

            // then
            expect(component).toMatchSnapshot();
        });
    })
})