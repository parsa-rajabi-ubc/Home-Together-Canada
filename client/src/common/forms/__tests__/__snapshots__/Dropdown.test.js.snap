// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Snapshot test should render correctly regardless of properties 1`] = `
<div
  className="relative inline-block text-left"
>
  <div
    className="inline-flex justify-center w-full rounded-md border border-gray-300 shadow-sm px-4 py-2 mx-3 bg-white text-sm font-medium text-gray-700 hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-indigo-500"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <span
      className="  rounded-md  bg-white ring-1 ring-black ring-opacity-25"
    >
      <p
        aria-expanded="true"
        aria-haspopup="true"
        className=""
        id="options-menu"
      >
        testTitleString1
         â†“ 
      </p>
    </span>
  </div>
</div>
`;
